File: Packages/Yage/Sources/Yage/Models/Hotspots.swift
import Schwifty
import SwiftUI

let boundsThickness: CGFloat = 2

// MARK: - Hotspots

public enum Hotspot: String, CaseIterable {
    case topBound
    case leftBound
    case rightBound
    case bottomBound
}

// MARK: - Default Hotspots

extension World {
    func hotspotEntities() -> [Entity] {
        [
            topBound(),
            bottomBound(),
            leftBound(),
            rightBound()
        ]
    }
}

// MARK: - Description

extension Hotspot: CustomStringConvertible {
    public var description: String { rawValue }
}

// MARK: - Collisions

public extension Collision {
    var hotspot: Hotspot? {
        Hotspot.allCases.first { $0.rawValue == bodyId }
    }
}

public extension Collisions {
    func contains(_ hotspot: Hotspot) -> Bool {
        first { $0.hotspot == hotspot } != nil
    }

    func contains(anyOf hotspots: [Hotspot]) -> Bool {
        hotspots.first { contains($0) } != nil
    }
}

// MARK: - Entities

private extension World {
    func bottomBound() -> Entity {
        let entity = Entity(
            species: .hotspot,
            id: Hotspot.bottomBound.rawValue,
            frame: CGRect(
                x: 0,
                y: bounds.height,
                width: bounds.width,
                height: boundsThickness
            ),
            in: bounds
        )
        entity.isStatic = true
        return entity
    }

    func topBound() -> Entity {
        let entity = Entity(
            species: .hotspot,
            id: Hotspot.topBound.rawValue,
            frame: CGRect(
                x: 0,
                y: 0,
                width: bounds.width,
                height: boundsThickness
            ),
            in: bounds
        )
        entity.isStatic = true
        return entity
    }

    func leftBound() -> Entity {
        let entity = Entity(
            species: .hotspot,
            id: Hotspot.leftBound.rawValue,
            frame: CGRect(
                x: 0,
                y: 0,
                width: boundsThickness,
                height: bounds.height
            ),
            in: bounds
        )
        entity.isStatic = true
        return entity
    }

    func rightBound() -> Entity {
        let entity = Entity(
            species: .hotspot,
            id: Hotspot.rightBound.rawValue,
            frame: CGRect(
                x: bounds.width,
                y: 0,
                width: boundsThickness,
                height: bounds.height
            ),
            in: bounds
        )
        entity.isStatic = true
        return entity
    }
}

File: Sources/windows/yage/models/hotspots.py
from enum import Enum
from yage.models.entity import Entity
from yage.models.species import SPECIES_HOTSPOT
from yage.utils.geometry import Rect

BOUNDS_THICKNESS = 2


class Hotspot(Enum):
    TOP_BOUND = 'top_bound'
    LEFT_BOUND = 'left_bound'
    RIGHT_BOUND = 'right_bound'
    BOTTOM_BOUND = 'bottom_bound'

    @classmethod
    def build_bottom_bound(cls, world) -> Entity:
        entity = Entity(
            SPECIES_HOTSPOT,
            Hotspot.BOTTOM_BOUND.value,
            Rect(
                0,
                world.bounds.height,
                world.bounds.width,
                BOUNDS_THICKNESS
            ),
            world
        )
        entity.is_static = True
        return entity

    @classmethod
    def build_top_bound(cls, world) -> Entity:
        entity = Entity(
            SPECIES_HOTSPOT,
            Hotspot.TOP_BOUND.value,
            Rect(0, 0, world.bounds.width, BOUNDS_THICKNESS),
            world
        )
        entity.is_static = True
        return entity

    @classmethod
    def build_left_bound(cls, world) -> Entity:
        entity = Entity(
            SPECIES_HOTSPOT,
            Hotspot.LEFT_BOUND.value,
            Rect(
                0,
                0,
                BOUNDS_THICKNESS,
                world.bounds.height
            ),
            world
        )
        entity.is_static = True
        return entity

    @classmethod
    def build_right_bound(cls, world) -> Entity:
        entity = Entity(
            SPECIES_HOTSPOT,
            Hotspot.RIGHT_BOUND.value,
            Rect(
                world.bounds.width,
                0,
                BOUNDS_THICKNESS,
                world.bounds.height
            ),
            world
        )
        entity.is_static = True
        return entity
